README.txt
CPSBayes
========
$Id$

CPSBayes is a tool to perform naive bayesian inference based on the BayesCore
package.

Data storage
------------

CPSBayes provides a CMF tool that holds the connection parameters to an
external SQL storage server (MySQL, PostgreSQL or SQLite).

Sample setups:

- MySQL: mysql://login:password@myhost/database_name
- PostgreSQL: postgres://user@host/database?debug=&cache;=
- SQLite: sqlite:///full/path/to/database

A ZODB storage is also provided and requires no external adapter nor setup.

ZMI screen
-----------

Two ZMI tabs are provided:

 - storage setup
 - categories definitions (listing, adding, deleting)

API
---

The CPSBayes tool defines the following methods:

- getCategoryList() : list existing categories
- addCategory(name, label='', description=''): add a new category
- delCategory(name) : delete a category
- learn(data, category): learn to guess `category` for the given `data`
- guess(data): return a sorted list of tuple (category_name, probability)

Sample usage
------------

Example (from Reverend)::

  >>> bayes_tool = self.portal.portal_bayes
  >>> bayes_tool.learn('le la les du un une je il elle de en', 'french')
  >>> bayes_tool.learn('der die das ein eine', 'german')
  >>> bayes_tool.learn('el uno una las de la en', 'spanish')
  >>> bayes_tool.learn('the rain in spain falls mainly on the plain', 'english')
  >>> bayes_tool.learn('the it she he they them are were to', 'english')
  >>> res = bayes_tool.guess('they went to el cantina')
  [('english', 0.9..), ('spanish', 0.9..)]
  >>> res = bayes_tool.guess('they were flying planes')
  [('english', 0.9..)]



CHANGES
Requires:
~~~~~~~~~
-
New features:
~~~~~~~~~~~~~
-
Bug fixes:
~~~~~~~~~~
-
New internal features:
~~~~~~~~~~~~~~~~~~~~~~
- added high level lockers


interfaces.py
# -*- encoding: iso-8859-15 -*-
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Author: Tarek Ziadé <tz@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
# $Id$

""" CPSBayes interfaces.
"""
from zope.interface import Interface

class ICPSBayes(Interface):
    """ CPSBayes
    """


exportimport.py
# -*- encoding: iso-8859-15 -*-
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# Author: Tarek Ziadé <tz@nuxeo.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
# $Id$
""" CPSBayes XML Adapter.
"""
from Acquisition import aq_base
from zope.component import adapts
from zope.interface import implements

from Products.CMFCore.utils import getToolByName
from Products.GenericSetup.utils import exportObjects
from Products.GenericSetup.utils import importObjects
from Products.GenericSetup.utils import XMLAdapterBase
from Products.GenericSetup.utils import ObjectManagerHelpers
from Products.GenericSetup.interfaces import IBody
from Products.GenericSetup.interfaces import ISetupEnviron

from Products.CPSBayes.interfaces import ICPSBayes

_marker = object()

TOOL = 'portal_bayes'
NAME = 'cpsbayes'

def exportCPSBayes(context):
    """Export user folder configuration as a set of XML files.

    Does not export the users themselves.
    """
    site = context.getSite()
    if getattr(aq_base(site), TOOL, None) is None:
        logger = context.getLogger(NAME)
        logger.info("Nothing to export.")
        return
    tool = getToolByName(site, TOOL)
    exportObjects(tool, '', context)

def importCPSBayes(context):
    """Import user folder configuration from XML files.
    """
    site = context.getSite()
    if getattr(aq_base(site), TOOL, None) is None:
        logger = context.getLogger(NAME)
        logger.info("Cannot import into missing acl_users.")
        return
    tool = getToolByName(site, TOOL)
    importObjects(tool, '', context)


class CPSBayesXMLAdapter(XMLAdapterBase, ObjectManagerHelpers):
    """XML importer and exporter for Standard User Folder.
    """

    adapts(ICPSBayes, ISetupEnviron)
    implements(IBody)

    _LOGGER_ID = NAME
    name = NAME

    def _exportNode(self):
        """Export the object as a DOM node.
        """
        node = self._getObjectNode('object')
        node.appendChild(self._extractCPSBayesProperties())
        self._logger.info("CPSBayes exported.")
        return node

    def _importNode(self, node):
        """Import the object from the DOM node.
        """
        meta_type = str(node.getAttribute('meta_type'))
        if meta_type != self.context.meta_type:
            self._logger.error("Cannot import %r into %r." %
                               (meta_type, self.context.meta_type))
            return
        if self.environ.shouldPurge():
            self._purgeCPSBayesProperties()
        self._initCPSBayesProperties(node)
        self._logger.info("CPSBayes imported.")

    node = property(_exportNode, _importNode)
    elements = ('backend_type', 'backend_parameters')

    def _extractCPSBayesProperties(self):
        portal_bayes = self.context
        fragment = self._doc.createDocumentFragment()

        for element in self.elements:
            child = self._doc.createElement('property')
            child.setAttribute('name', element)
            value = getattr(portal_bayes, element)
            text = self._doc.createTextNode(str(value))
            child.appendChild(text)
            fragment.appendChild(child)
        return fragment

    def _purgeCPSBayesProperties(self):
        return

    def _initCPSBayesProperties(self, node):
        portal_bayes = self.context

        for child in node.childNodes:
            if child.nodeName != 'property':
                continue
            name = child.getAttribute('name')
            value = self._getNodeText(child)
            if name in self.elements:
                setattr(portal_bayes, name, value)
